compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/IsolationAnalyzer.java

```diff
@@ -1411,16 +1413,83 @@ public void visit(BLangInvocation.BLangActionInvocation actionInvocationExpr) {
             return;
         }
 
-        markDependsOnIsolationNonInferableConstructs();
-        inferredIsolated = false;
+        boolean isWorker = actionInvocationExpr.functionPointerInvocation;
+        if (isInIsolatedFunction(env)) {
+            if (checkStrandAnnotationExists(actionInvocationExpr.annAttachments, true, isWorker)) {
+                return;
+            }
+
+            if (!isWorker && !isValidIsolatedAsyncInvocation(actionInvocationExpr)) {
+                return;
+            }
+
+            analyzeInvocation(actionInvocationExpr);
+            return;
+        }
 
-        if (actionInvocationExpr.functionPointerInvocation) {
+        if (checkStrandAnnotationExists(actionInvocationExpr.annAttachments, false, isWorker)) {
+            markDependsOnIsolationNonInferableConstructs();
+            inferredIsolated = false;
             return;
         }
 
-        if (isInIsolatedFunction(env.enclInvokable)) {
-            dlog.error(actionInvocationExpr.pos, DiagnosticErrorCode.INVALID_ASYNC_INVOCATION_IN_ISOLATED_FUNCTION);
+        analyzeInvocation(actionInvocationExpr);
+    }
+
+    private boolean checkStrandAnnotationExists(List<BLangAnnotationAttachment> attachments, boolean inIsolatedFunc,
+                                                boolean isWorker) {
+        BAnnotationSymbol strandAnnotSymbol = symResolver.getStrandAnnotationSymbol();
+        String actionInvocation = isWorker ? "worker declaration" : "start action";
+        for (BLangAnnotationAttachment attachment : attachments) {
+            if (attachment.annotationSymbol == strandAnnotSymbol) {
+                if (inIsolatedFunc) {
+                    dlog.error(attachment.pos, DiagnosticErrorCode.INVALID_STRAND_ANNOTATION_IN_ISOLATED_FUNCTION,
```

Comment:
The error messages currently seem to say `ballerina/lang.annotations:0.0.0:strand`. The annotations module is an implementation detail, and we shouldn't include it in the error. We can just say `'strand'`.

Even though the strand annotation is also implementation-dependent it works like a builtin definition. Similar to builtin types.